File: IFATCCompanion/Models/FlightPhase.swift
import Foundation

/// Physical phase of flight inferred from aircraft state by `PhaseDetector`,
/// and produced directly by `MockSimulatorFeed`.
enum FlightPhase: String, CaseIterable, Codable, Identifiable {
    case preflight
    case taxiOut
    case takeoff
    case initialClimb
    case climb
    case cruise
    case descent
    case approach
    case landing
    case taxiIn
    case parked
    case unknown

    var id: String { rawValue }

    var title: String {
        switch self {
        case .preflight: return "Preflight"
        case .taxiOut: return "Taxi Out"
        case .takeoff: return "Takeoff"
        case .initialClimb: return "Initial Climb"
        case .climb: return "Climb"
        case .cruise: return "Cruise"
        case .descent: return "Descent"
        case .approach: return "Approach"
        case .landing: return "Landing"
        case .taxiIn: return "Taxi In"
        case .parked: return "Parked"
        case .unknown: return "Unknown"
        }
    }

    /// Whether the aircraft is expected to be on the ground in this phase.
    var isGround: Bool {
        switch self {
        case .preflight, .taxiOut, .taxiIn, .parked: return true
        default: return false
        }
    }

    /// Demo ordering used by the mock feed "advance phase" control.
    static let demoSequence: [FlightPhase] = [
        .preflight, .taxiOut, .takeoff, .initialClimb, .climb,
        .cruise, .descent, .approach, .landing, .taxiIn, .parked
    ]
}

File: Android/core/src/main/kotlin/com/h3consultingpartners/ifatccompanion/core/model/FlightPhase.kt
package com.h3consultingpartners.ifatccompanion.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Physical phase of flight inferred from aircraft state by `PhaseDetector`, and
 * produced directly by `MockSimulatorFeed`.
 *
 * Ported from `IFATCCompanion/Models/FlightPhase.swift`.
 */
@Serializable
enum class FlightPhase(val rawValue: String) {
    @SerialName("preflight") PREFLIGHT("preflight"),
    @SerialName("taxiOut") TAXI_OUT("taxiOut"),
    @SerialName("takeoff") TAKEOFF("takeoff"),
    @SerialName("initialClimb") INITIAL_CLIMB("initialClimb"),
    @SerialName("climb") CLIMB("climb"),
    @SerialName("cruise") CRUISE("cruise"),
    @SerialName("descent") DESCENT("descent"),
    @SerialName("approach") APPROACH("approach"),
    @SerialName("landing") LANDING("landing"),
    @SerialName("taxiIn") TAXI_IN("taxiIn"),
    @SerialName("parked") PARKED("parked"),
    @SerialName("unknown") UNKNOWN("unknown"),
    ;

    val id: String get() = rawValue

    val title: String
        get() = when (this) {
            PREFLIGHT -> "Preflight"
            TAXI_OUT -> "Taxi Out"
            TAKEOFF -> "Takeoff"
            INITIAL_CLIMB -> "Initial Climb"
            CLIMB -> "Climb"
            CRUISE -> "Cruise"
            DESCENT -> "Descent"
            APPROACH -> "Approach"
            LANDING -> "Landing"
            TAXI_IN -> "Taxi In"
            PARKED -> "Parked"
            UNKNOWN -> "Unknown"
        }

    /** Whether the aircraft is expected to be on the ground in this phase. */
    val isGround: Boolean
        get() = when (this) {
            PREFLIGHT, TAXI_OUT, TAXI_IN, PARKED -> true
            else -> false
        }

    companion object {
        /** Demo ordering used by the mock feed "advance phase" control. */
        val demoSequence: List<FlightPhase> = listOf(
            PREFLIGHT, TAXI_OUT, TAKEOFF, INITIAL_CLIMB, CLIMB,
            CRUISE, DESCENT, APPROACH, LANDING, TAXI_IN, PARKED,
        )

        fun fromRawValue(raw: String): FlightPhase? = entries.firstOrNull { it.rawValue == raw }
    }
}
